File: tests_MATLAB/test4.m
% Energy test

addpath('../mole_MATLAB')

% Parameters
k = 4;
a = -5;
b = 5;
m = 500;
grid = linspace(a, b, m);
dx = grid(2) - grid(1);
tol = 1e-6;

% Laplacian
L = lap(k, m - 2, dx);

% Transform grid using an anonymous function
grid = arrayfun(@(x) x^2, grid);

% Potential matrix V
V = spdiags(grid', 0, m, m);  % Transpose grid to match dimension

% Hamiltonian H
H = -0.5 * L + V;

% Eigenvalue computation
eigval = eig(full(H));  % Convert sparse matrix to full for eig function
eigval = sort(eigval);

% Expected eigenvalues
expected = [1, 3, 5, 7, 9];

% Test for failure
failed = false;
for i = 1:length(expected)
    if (abs(real(eigval(i) / eigval(1)) - expected(i)) > tol)
        fprintf(2, 'Test FAILED!\n');
        failed = true;
    end
end

if ~failed
    fprintf(1, 'Test PASSED!\n');
end

File: tests_C++/test4.cpp
/**
 * Energy test
 */

#include "mole.h"
#include <algorithm>
#include <iostream>

int main() {

  int k = 4;
  Real a = -5;
  Real b = 5;
  int m = 500;
  vec grid = linspace(a, b, m);
  Real dx = grid(1) - grid(0);
  Real tol = 1e-10;

  Laplacian L(k, m - 2, dx);

  std::transform(grid.begin(), grid.end(), grid.begin(),
                 [](Real x) { return x * x; });

  sp_mat V(m, m);
  V.diag(0) = grid;

  sp_mat H = -0.5 * (sp_mat)L + V;

  cx_vec eigval;
  eig_gen(eigval, (mat)H);

  eigval = sort(eigval);

  vec expected{1, 3, 5, 7, 9};

  bool failed = false;
  for (int i = 0; i < expected.size(); ++i)
    if (std::norm(real(eigval(i) / eigval(0)) - expected(i)) > tol) {
      cout << "\033[1;31mTest FAILED!\033[0m\n";
      failed = true;
    }

  if (!failed)
    cout << "\033[1;32mTest PASSED!\033[0m\n";

  return 0;
}
